homeassistant/components/intent/timers.py

```diff
@@ -0,0 +1,772 @@
+
+    async def add_time(self, timer_id: str, seconds: int) -> None:
+        """Add time to a timer."""
+        if seconds == 0:
+            return
+
+        timer = self.timers.get(timer_id)
+        if timer is None:
+            return
+
+        timer.seconds = max(0, timer.seconds_left + seconds)
+        timer.updated_at = time.monotonic_ns()
+        if not timer.is_paused:
+            timer.task.cancel()
+            timer.task = self.hass.async_create_background_task(
+                self._wait_for_timer(timer_id, timer.seconds, timer.updated_at),
+                name=f"Timer {timer_id}",
+            )
+
+        self.hass.bus.async_fire(EVENT_INTENT_TIMER_UPDATED, timer.to_event())
+
+        if seconds > 0:
+            log_verb = "increased"
+            log_seconds = seconds
+        else:
+            log_verb = "decrease"
+            log_seconds = -seconds
+
+        _LOGGER.debug(
+            "Timer %s by %s second(s): id=%s, name=%s, seconds_left=%s, device_id=%s",
+            log_verb,
+            log_seconds,
+            timer_id,
+            timer.name,
+            timer.seconds_left,
+            timer.device_id,
+        )
+
+    async def remove_time(self, timer_id: str, seconds: int) -> None:
+        """Remove time from a timer."""
+        await self.add_time(timer_id, -seconds)
+
+    async def pause_timer(self, timer_id: str) -> None:
+        """Pauses a timer."""
+        timer = self.timers.get(timer_id)
+        if (timer is None) or timer.is_paused:
+            return
+
+        timer.seconds = timer.seconds_left
+        timer.updated_at = time.monotonic_ns()
+        timer.is_paused = True
+        timer.task.cancel()
+
+        self.hass.bus.async_fire(EVENT_INTENT_TIMER_UPDATED, timer.to_event())
+
+        _LOGGER.debug(
+            "Timer paused: id=%s, name=%s, seconds_left=%s, device_id=%s",
+            timer_id,
+            timer.name,
+            timer.seconds_left,
+            timer.device_id,
+        )
+
+    async def unpause_timer(self, timer_id: str) -> None:
+        """Unpause a timer."""
+        timer = self.timers.get(timer_id)
+        if (timer is None) or (not timer.is_paused):
+            return
+
+        timer.is_paused = False
+        timer.updated_at = time.monotonic_ns()
+        timer.task = self.hass.async_create_background_task(
+            self._wait_for_timer(timer_id, timer.seconds_left, timer.updated_at),
+            name=f"Timer {timer.id}",
+        )
+
+        self.hass.bus.async_fire(EVENT_INTENT_TIMER_UPDATED, timer.to_event())
+
+        _LOGGER.debug(
+            "Timer unpaused: id=%s, name=%s, seconds_left=%s, device_id=%s",
+            timer_id,
+            timer.name,
+            timer.seconds_left,
+            timer.device_id,
+        )
+
+    async def _timer_finished(self, timer_id: str) -> None:
+        """Call event handlers when a timer finishes."""
+        timer = self.timers.pop(timer_id, None)
+        if timer is None:
```

Comment:
can this happen?

Reply:
I don't think so given we're not using threads. Removed.